### PR TITLE
[libclang/python] Refactor enum usage

### PR DESCRIPTION
**Pull request by DeinAlptraum**
Use Python's builtin enum class instead of writing our own.

This is preparation for passing a strict type check in PR #78114 , fixing 927 out of 1341 strict typing errors

### CODE REVIEW

File: clang/bindings/python/tests/cindex/test_enums.py
Comment (DeinAlptraum) [inline clang/bindings/python/tests/cindex/test_enums.py:47]: This test is effectively pointless now, since the enum class errors out as soon as the module is loaded in case there are duplicate enum variants

File: clang/bindings/python/clang/cindex.py
Comment (DeinAlptraum) [inline clang/bindings/python/clang/cindex.py:648]: The `Enum` class already provides a `name` attribute that follows the exact same format as we had before

File: clang/bindings/python/clang/cindex.py
Comment (DeinAlptraum) [inline clang/bindings/python/clang/cindex.py:629]: In hindsight, should the change of error message go into a separate PR? The one we had before, `Unknown template argument kind`, doesn't make much sense since this is the base class of not just `TemplateArgumentKind`.
If we don't consider changing error messages as a breaking change, then we could also remove the try-except part entirely since `Enum` already throws a meanginful error, looking like this: `ValueError: 703 is not a valid CursorKind`

File: clang/bindings/python/clang/cindex.py
Comment (DeinAlptraum) [inline clang/bindings/python/clang/cindex.py:633]: This `__repr__` was effectively unused before as each subclass defined its own. This change is necessary so that all subclasses still return exactly the same representation as they did before

**Review by DeinAlptraum — COMMENTED** [review]
There is also `TokenKind`: this one does not currently inherit from `BaseEnumeration` and is defined somewhat differently, having all its variants and their IDs as a dictionary in `enumerations.py`. This seems quite arbitrary to me, is there any reason it is done this way? Otherwise I would also move this to `cindex.py` as another subclass of `BaseEnumeration`

**Comment by DeinAlptraum** [comment]
@Endilll can I ask you for a review again?
As a next step towards the python-bindings strict typing PR, this one captures all the enum refactoring changes necessary towards that goal. Don't be scared by the LoC changed: 90% of that is just indentation changes :)

**Comment by DeinAlptraum** [comment]
@Endilll are you taking a look at this, and/or should I ask other reviewers?

**Review by Endilll — COMMENTED** [review]
LGTM

> There is also TokenKind: this one does not currently inherit from BaseEnumeration and is defined somewhat differently, having all its variants and their IDs as a dictionary in enumerations.py. This seems quite arbitrary to me, is there any reason it is done this way? Otherwise I would also move this to cindex.py as another subclass of BaseEnumeration

I agree that this seems arbitrary. I think you can proceed with refactoring `TokenKind`.

**Comment by Endilll** [comment]
> @Endilll are you taking a look at this, and/or should I ask other reviewers?

It's never a bad idea to add more reviewers, as long as they are relevant.

**Comment by DeinAlptraum** [comment]
@Endilll done with the TokenKind refactoring, so this is ready to go from my side.

Regarding reviewers, can you recommend any others? Since I assume I'll open a couple more PRs for the Python bindings after this

**Comment by Endilll** [comment]
Also worth pinging @linux4life798

**Comment by DeinAlptraum** [comment]
\*Ping*

@boomanaiden154 @linux4life798 a review/feedback would be appreciated!

File: clang/bindings/python/tests/cindex/test_token_kind.py
Comment (Endilll) [inline clang/bindings/python/tests/cindex/test_token_kind.py:13]: Can't we keep those tests?

Comment (DeinAlptraum) [reply clang/bindings/python/tests/cindex/test_token_kind.py:13]: thNot really:
- `test_constructor`: this tests the ability to add enum variants "on the fly", which is not possible with the Python stdlib's `Enum` class
- `test_bad_register`: the `Enum` class tests for this on initialization, so if you define a duplicate `Enum` variant, you will get a failure on import already. This makes the test both a) effectively covered by that and b) you couldn't import the module to run such a test anyway in a case where there's a duplicate
- `test_unknown_value`: this is covered in `test_from_id` in `test_enums.py` since I added the `TokenKind` class to the test there

Comment (Endilll) [reply clang/bindings/python/tests/cindex/test_token_kind.py:13]: > test_constructor: this tests the ability to add enum variants "on the fly", which is not possible with the Python stdlib's Enum class

Do we have a test that this is not possible anymore?

Comment (DeinAlptraum) [reply clang/bindings/python/tests/cindex/test_token_kind.py:13]: So the existing "custom" enum implementation included a sort of constructor, as we see in this test, that takes a number and name and then adds that as a new variant to the existing enum. No such constructor exists for the builtin `Enum` class, so there isn't really an equivalent for this test under the new implementation.
We could test that it is impossible for the user to add their own variants dynamically at runtime by doing something like `TokenKind.NEW_VARIANT = 6`, but that doesn't really seem like our problem, and was also possible under the old implementation

Your comment did however raise another question for me: while we've covered that there are no duplicate names with different IDs, we haven't covered that there are no duplicate IDs with different names, and this is indeed possible in the enum class, and not covered by our current tests so I just added one.

**Comment by DeinAlptraum** [comment]
@Endilll do you think this needs a second review?
If not, could you merge this? Since I don't have commit access

**Comment by DeinAlptraum** [comment]
*Ping\*
@Endilll how should we proceed with this?

File: clang/bindings/python/clang/cindex.py
Comment (Endilll) [reply clang/bindings/python/clang/cindex.py:589]: Changing error messages is not breaking, because we don't promise their stability anywhere as far as I know. Given that those are error messages about enums, I think such a change would belong to this PR if you want to move in this direction.

On a separate note, do we even need `from None` here? I think it's implicit anyway.

**Review by Endilll — COMMENTED** [review]
Sorry for not getting back to you earlier.

File: clang/bindings/python/clang/cindex.py
Comment (DeinAlptraum) [reply clang/bindings/python/clang/cindex.py]: Thank you for the feedback! I've removed the `try`-`except` block now so we're just using the `Enum` error message

Irrelevant now, but regarding `from None`:
This effectively removes the previous exception context. Since we are re-raising an exception here, we would get both the exception we raised ourselves _and_  the exception that we caught here.

E.g. currently we get 
```
>>> cindex.TokenKind.from_id(5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/shared/programming/repos/llvm-project/clang/bindings/python/clang/cindex.py", line 589, in from_id
    raise ValueError("Unknown %s %d" % (cls.__name__, id)) from None
ValueError: Unknown TokenKind 5
```

and without the `from None` we would get
```
>>> cindex.TokenKind.from_id(5)
Traceback (most recent call last):
  File "/data/shared/programming/repos/llvm-project/clang/bindings/python/clang/cindex.py", line 587, in from_id
    return cls(id)
           ^^^^^^^
  File "/usr/lib/python3.12/enum.py", line 757, in __call__
    return cls.__new__(cls, value)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/enum.py", line 1171, in __new__
    raise ve_exc
ValueError: 5 is not a valid TokenKind

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/shared/programming/repos/llvm-project/clang/bindings/python/clang/cindex.py", line 589, in from_id
    raise ValueError("Unknown %s %d" % (cls.__name__, id))
ValueError: Unknown TokenKind 5
```

**Comment by DeinAlptraum** [comment]
@Endilll do you think this can be merged now?

**Comment by Endilll** [comment]
> @Endilll do you think this can be merged now?

After CI passes.

**Comment by Endilll** [comment]
@DeinAlptraum This is also worth a release note, I think.

**Comment by DeinAlptraum** [comment]
Hmm what do you think should be mentioned in a release note @Endilll ?
Since this is just refactoring, if I did everything correctly, users of the bindings should not notice any difference (except for different error messages). So just mention refactoring of enums in the bindings?

**Comment by Endilll** [comment]
The refactoring itself is worth mentioning in potentially breaking changes, because this covers a significant surface of our Python API.

**Comment by DeinAlptraum** [comment]
I added a release note to the `Clang Python Bindings Potentially Breaking Changes` section

File: clang/docs/ReleaseNotes.rst
Comment (Endilll) [inline clang/docs/ReleaseNotes.rst:145]:
```suggestion
- Refactored enum implementation, switching to the standard library `Enum` type.
```

**Review by Endilll — APPROVED** [review]

**Pull request merged**
